### PR TITLE
feat: add E2E tests for cases that peers going offline

### PR DESCRIPTION
**Pull request by BruceAko**
## Description

Verifies that Scheduler can clear offline peer metadata when the host exits normally/abnormally by following steps:

1. Get the number of all hosts in the scheduler.
2. After a normal exit of a host, verify that the number of hosts in the scheduler decreases by 1.
3. After an abnormal exit of a host, verify that the number of hosts in the scheduler decreases by 1 after twice the announce host interval.

## Related Issue

close https://github.com/dragonflyoss/Dragonfly2/issues/3523

## Motivation and Context

Earlier we have added functionality so that hosts send `LeaveHost()` requests to the Scheduler on normal exit and hosts on abnormal exit are also sensed by the Scheduler GC after a set amount of time and the metadata is cleared. Now, we are going to add E2E tests to verify that this functionality is correct.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: .github/workflows/compatibility-e2e-v2.yml [hunk 1]
```diff
@@ -45,6 +45,7 @@ jobs:
             image: client
             image-tag: v0.1.100
             chart-name: seed-client
+        skip: "Clients\sgo\soffline\snormally\sand\sabnormally"
```
Comment (gaius-qi) [inline .github/workflows/compatibility-e2e-v2.yml:48]: Remove `\s`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: .github/workflows/compatibility-e2e-v2.yml (hunk 1 above)
Comment (BruceAko) [reply .github/workflows/compatibility-e2e-v2.yml]: Resolved.

**Comment by BruceAko** [comment]
@gaius-qi All problems have been resolved, please take a look.

File: .github/workflows/e2e-v2-nydus.yml [hunk 2]
```diff
@@ -18,7 +18,7 @@ env:
   KIND_VERSION: v0.12.0
   CONTAINERD_VERSION: v1.5.2
   NERDCTL_VER: 0.22.2
-  KIND_CONFIG_PATH: test/testdata/kind/config-v2.yaml
+  KIND_CONFIG_PATH: test/testdata/kind/config-v2-nydus.yaml
```
Comment (gaius-qi) [inline .github/workflows/e2e-v2-nydus.yml:21]: Why create configuration?

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: test/testdata/kind/config-v2.yaml [hunk 3]
```diff
@@ -12,6 +12,18 @@ nodes:
...
+        protocol: TCP
+    extraMounts:
+      - hostPath: ./test/testdata/containerd/config-v2.toml
+        containerPath: /etc/containerd/config.toml
+      - hostPath: /tmp/artifact
+        containerPath: /tmp/artifact
+      - hostPath: /dev/fuse
+        containerPath: /dev/fuse
+  - role: worker
+    image: kindest/node:v1.23.4
```
Comment (gaius-qi) [inline test/testdata/kind/config-v2.yaml:26]: Remove `/dev/fuse`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e-v2-nydus.yml (hunk 2 above)
Comment (BruceAko) [reply .github/workflows/e2e-v2-nydus.yml]: Removed now. 
At the beginning, the workflow was set up with two workers and one control plane, but now it's changed to one worker and one control plane, and it doesn't report insufficient disk space error when kind load image. I've changed back to using the same config.

File: .github/workflows/e2e-v2-nydus.yml [hunk 4]
```diff
@@ -103,6 +103,11 @@ jobs:
           kind load docker-image dragonflyoss/client:latest
           kind load docker-image dragonflyoss/dfinit:latest
 
+      - name: Remove and add taints to node
```
Comment (gaius-qi) [inline .github/workflows/e2e-v2-nydus.yml:106]: Remove.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e-v2.yml [hunk 5]
```diff
@@ -120,6 +120,11 @@ jobs:
           kind load docker-image dragonflyoss/client:latest
           kind load docker-image dragonflyoss/dfinit:latest
 
+      - name: Remove and add taints to node
```
Comment (gaius-qi) [inline .github/workflows/e2e-v2.yml:123]: Remove.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/rpc/scheduler/client/client_v2.go [hunk 6]
```diff
@@ -145,6 +146,9 @@ type V2 interface {
 	// AnnounceHost announces host to scheduler.
 	AnnounceHost(context.Context, *schedulerv2.AnnounceHostRequest, ...grpc.CallOption) error
 
+	// ListHosts lists hosts in scheduler.
+	ListHosts(ctx context.Context, opts ...grpc.CallOption) (*schedulerv2.ListHostsResponse, error)
```
Comment (gaius-qi) [inline pkg/rpc/scheduler/client/client_v2.go:150]: Add `taskID` params.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/service/service_v2_test.go [hunk 7]
```diff
@@ -882,6 +882,115 @@ func TestServiceV2_AnnounceHost(t *testing.T) {
...
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctl := gomock.NewController(t)
+			defer ctl.Finish()
+			scheduling := schedulingmocks.NewMockScheduling(ctl)
+			res := resource.NewMockResource(ctl)
+			dynconfig := configmocks.NewMockDynconfigInterface(ctl)
+			storage := storagemocks.NewMockStorage(ctl)
+			hostManager := resource.NewMockHostManager(ctl)
+			host := resource.NewHost(
```
Comment (gaius-qi) [inline scheduler/service/service_v2_test.go:976]: Use option pattern.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: test/e2e/v2/leave_host_test.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	. "github.com/onsi/ginkgo/v2" //nolint
+	. "github.com/onsi/gomega"    //nolint
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	schedulerclient "d7y.io/dragonfly/v2/pkg/rpc/scheduler/client"
+	"d7y.io/dragonfly/v2/test/e2e/v2/util"
+)
+
+var _ = Describe("Clients go offline normally and abnormally", func() {
```
Comment (gaius-qi) [inline test/e2e/v2/leave_host_test.go:34]: `Client Leaving`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/service/service_v2.go [hunk 9]
```diff
@@ -699,6 +699,84 @@ func (v *V2) AnnounceHost(ctx context.Context, req *schedulerv2.AnnounceHostRequ
 	return nil
 }
 
+// ListHosts lists hosts in scheduler.
+func (v *V2) ListHosts(ctx context.Context) (*schedulerv2.ListHostsResponse, error) {
+	hosts := v.resource.HostManager().LoadAll()
+
+	resHosts := make([]*commonv2.Host, len(hosts))
```
Comment (gaius-qi) [inline scheduler/service/service_v2.go:706]: `respHosts`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: test/testdata/kind/config-v2.yaml (hunk 3 above)
Comment (BruceAko) [reply test/testdata/kind/config-v2.yaml]: Resolved.

File: .github/workflows/e2e-v2-nydus.yml (hunk 4 above)
Comment (BruceAko) [reply .github/workflows/e2e-v2-nydus.yml]: Resolved.

File: pkg/rpc/scheduler/client/client_v2.go (hunk 6 above)
Comment (BruceAko) [reply pkg/rpc/scheduler/client/client_v2.go]: Resolved.

File: scheduler/service/service_v2_test.go (hunk 7 above)
Comment (BruceAko) [reply scheduler/service/service_v2_test.go:976]: Resolved.

File: test/e2e/v2/leave_host_test.go (hunk 8 above)
Comment (BruceAko) [reply test/e2e/v2/leave_host_test.go]: Resolved.

File: scheduler/service/service_v2.go (hunk 9 above)
Comment (BruceAko) [reply scheduler/service/service_v2.go]: Resolved.

**Comment by BruceAko** [comment]
@gaius-qi  Please take a look, thanks :)

File: test/testdata/charts/config-v2.yaml [hunk 10]
```diff
@@ -111,6 +114,10 @@ client:
     limits:
       cpu: "2"
       memory: "4Gi"
+  tolerations:
```
Comment (gaius-qi) [inline test/testdata/charts/config-v2.yaml:117]: Add comment.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Comment by BruceAko** [comment]
@gaius-qi PTAL, thanks

File: test/e2e/v2/leave_host_test.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+			// wait for host gc
+			time.Sleep(2 * time.Minute)
+			Expect(getHostCountFromScheduler(schedulerClient)).To(Equal(hostCount - 1))
+
+			// remove taint in master node
+			out, err = util.KubeCtlCommand("-n", util.DragonflyNamespace, "taint", "nodes", "kind-control-plane", "master:NoSchedule-").CombinedOutput()
+			fmt.Println(string(out))
+			Expect(err).NotTo(HaveOccurred())
+			time.Sleep(30 * time.Second)
```
Comment (gaius-qi) [inline test/e2e/v2/leave_host_test.go:106]: Why sleep?

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: test/e2e/v2/util/exec.go [hunk 12]
```diff
@@ -103,37 +103,85 @@ func KubeCtlCopyCommand(ns, pod, source, target string) *exec.Cmd {
...
-	out, err := KubeCtlCommand("-n", DragonflyNamespace, "get", "pod", "-l", "component=client",
-		"-o", fmt.Sprintf("jsonpath='{range .items[0]}{.metadata.name}{end}'")).CombinedOutput()
+	podName, err := GetClientPodNameInWorker()
+	if err != nil {
+		return nil, err
+	}
+	return NewPodExec(DragonflyNamespace, podName, "client"), nil
+}
+
+func ClientExecAll() ([]*PodExec, error) {
```
Comment (gaius-qi) [inline test/e2e/v2/util/exec.go:113]: `AllClientExec`

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: test/e2e/v2/leave_host_test.go (hunk 11 above)
Comment (BruceAko) [reply test/e2e/v2/leave_host_test.go]: Removed.

File: test/e2e/v2/util/exec.go (hunk 12 above)
Comment (BruceAko) [reply test/e2e/v2/util/exec.go]: Resolved.

File: test/e2e/v2/leave_host_test.go [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	Context("normally", func() {
+		It("number of hosts should be ok", Label("host", "leave"), func() {
+			// create scheduler GRPC client
+			grpcCredentials := insecure.NewCredentials()
+			schedulerClient, err := schedulerclient.GetV2ByAddr(context.Background(), ":8002", grpc.WithTransportCredentials(grpcCredentials))
+			Expect(err).NotTo(HaveOccurred())
+
+			// get host count
+			hostCount := util.Servers[util.SeedClientServerName].Replicas + util.Servers[util.ClientServerName].Replicas
+			time.Sleep(10 * time.Minute)
```
Comment (gaius-qi) [inline test/e2e/v2/leave_host_test.go:44]: Move to `BeforeSuite`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: test/e2e/v2/leave_host_test.go [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+var _ = Describe("Clients Leaving", func() {
+	Context("normally", func() {
+		It("number of hosts should be ok", Label("host", "leave"), func() {
+			// create scheduler GRPC client
+			grpcCredentials := insecure.NewCredentials()
+			schedulerClient, err := schedulerclient.GetV2ByAddr(context.Background(), ":8002", grpc.WithTransportCredentials(grpcCredentials))
+			Expect(err).NotTo(HaveOccurred())
+
+			// get host count
+			hostCount := util.Servers[util.SeedClientServerName].Replicas + util.Servers[util.ClientServerName].Replicas
```
Comment (gaius-qi) [inline test/e2e/v2/leave_host_test.go:43]: Remove `util.SeedClientServerName].Replicas`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Comment by BruceAko** [comment]
All problems resolved, please take a look, thanks. @gaius-qi

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**